handling-null-values-in-avro-and-protobuf/kafka/README.md

```diff
@@ -0,0 +1,195 @@
+            avroConsumerRecords.forEach(avroConsumerRecord -> {
+                PurchaseAvro avroPurchase = avroConsumerRecord.value();
+                System.out.print("Purchase details consumed from topic with Avro schema { ");
+                System.out.printf("Customer: %s, ", avroPurchase.getCustomerId());
+                System.out.printf("Total Cost: %f, ", avroPurchase.getTotalCost());
+                System.out.printf("Item: %s } %n", avroPurchase.getItem());
+
+            });
+
+```
+
+Copy and paste the contents of `path to file/s on GitHub` into the `AvroConsumer.java` and `AvroProducer.java` files.
+
+Create a file at `handling-null-values/kafka/code/src/main/avro/purchase.avsc`. 
+
+Copy and paste the following into that file:
+
+```
+{
+  "type":"record",
+  "namespace": "io.confluent.developer.avro",
+  "name":"PurchaseAvro",
+  "fields": [
+    {"name": "item", "type": "string" },
+    {"name": "total_cost", "type": "double" },
+    {"name": "customer_id", "type": "string"}
+  ]
+}
+```
+
+If you run the code using `./gradlew runAvroProducer`, you will see that the producer "hangs" and does not produce events.
+
+For the "item" field in `purchase.avsc`, put:
+
+```
+{
+  "type":"record",
+  "namespace": "io.confluent.developer.avro",
+  "name":"PurchaseAvro",
+  "fields": [
+    {"name": "item", "type": ["string", "null"] },
+    {"name": "total_cost", "type": "double" },
+    {"name": "customer_id", "type": "string"}
+  ]
+}
+```
+When you run `./gradlew runAvroProducer` and furthermore, `./gradlew runAvroConsumer`, you'll see that the events with null items are produced and consumed successfully. 
+
+Next, copy and paste the contents of `path to file/s on GitHub` into the `ProtoConsumer.java` and `ProtoProducer.java` files.
+
+Create a file at `handling-null-values/kafka/code/src/main/proto/purchase.proto`. 
+
+Copy and paste the following into that file:
+
+```
+syntax = "proto3";
+
+package io.confluent.developer.proto;
+option java_outer_classname = "PurchaseProto";
+
+message Purchase {
+  string item = 1;
+  double total_cost = 2;
+  string customer_id = 3;
+}
+```
+
+Look at `ProtoProducerApp.java`, lines 76-77:
+
+```java
+        purchaseBuilder.setCustomerId("Customer Null")
+                .setTotalCost(random.nextDouble() * random.nextInt(100));
+``` 
+
+We can see that the developer who wrote this app 'forgot' to write the `setItem()` method that adds an item. This means that the value will be null. But when you run you run `./gradlew runProtoProducer` and `./gradlew runProtoConsumer` no errors will arise. That's because Protobuf automatically handles default values. 
+
+Now, if you _explicitly_ set the value of the item to null:
+
+
+```java
+        purchaseBuilder.setCustomerId("Customer Null").setItem(null)
+                .setTotalCost(random.nextDouble() * random.nextInt(100));
+``` 
+
+In this case, you'll receive a NullPointer error. 
```

Comment:
whats going to come next? the field being `oneof` where one is a boolean to signify null?

Reply:
Does `optional` do the trick? i.e. does everything just work if you do that? I'm looking at the guidance here:

>If not using optional would add complexity or ambiguity, then use optional primitive fields. Wrapper types must not be used going forward.

which makes me think we might want to avoid the wrapper pattern... I'm not sure why this is the guidance, e.g., will wrapper types eventually be deprecated?

I'm thinking that if `optional` works, go with that, and we'd also want to update [this](https://docs.confluent.io/cloud/current/sr/fundamentals/serdes-develop/serdes-protobuf.html#handling-null-values-with-wrappers) docs section to recommend `optional` rather than use a wrapper

wdyt?

Reply:
so with this protobuf definition:

```
message Purchase {
  optional string item = 1;
  double total_cost = 2;
  string customer_id = 3;
}
```

you still end up with a `NullPointerException` ¯\_(ツ)_/¯

but that's a good point about the guidance, I'll pass it on to docs.